// @wix/motion/dist/esm/library/entrance/ExpandIn.js
const SCALE_MAP = {
  soft: 0.8,
  medium: 0.6,
  hard: 0
};
const TRANSFORM_ORIGIN_MAP = {
  top: {
    x: 0,
    y: -0.5
  },
  'top-right': {
    x: 0.5,
    y: -0.5
  },
  right: {
    x: 0.5,
    y: 0
  },
  'bottom-right': {
    x: 0.5,
    y: 0.5
  },
  bottom: {
    x: 0,
    y: 0.5
  },
  'bottom-left': {
    x: -0.5,
    y: 0.5
  },
  left: {
    x: -0.5,
    y: 0
  },
  'top-left': {
    x: -0.5,
    y: -0.5
  },
  center: {
    x: 0,
    y: 0
  }
};
export default function create(options, dom) {
  const {
    power,
    initialScale = 0,
    direction = 'center'
  } = options.namedEffect;
  const easing = options.easing || 'cubicInOut';
  const scale = power && power in SCALE_MAP ? SCALE_MAP[power] : initialScale;
  const {
    x,
    y
  } = TRANSFORM_ORIGIN_MAP[direction];
  if (dom) {
    dom.measure(target => {
      if (!target) {
        return;
      }
      const {
        width,
        height
      } = target.getBoundingClientRect();
      dom.mutate(() => {
        target.style.setProperty('--motion-width', `${width}px`);
        target.style.setProperty('--motion-height', `${height}px`);
      });
    });
  }
  return [{
    ...options,
    easing: 'linear',
    keyframes: [{
      opacity: 0
    }, {
      opacity: 'var(--comp-opacity, 1)'
    }]
  }, {
    ...options,
    easing,
    keyframes: [{
      transform: `translateX(calc(var(--motion-width, 100%) * ${x})) translateY(calc(var(--motion-height, 100%) * ${y})) scale(${scale}) translateX(calc(var(--motion-width, 100%) * ${-x})) translateY(calc(var(--motion-height, 100%) * ${-y}))  rotate(var(--comp-rotate-z, 0deg))`
    }, {
      transform: `translateX(calc(var(--motion-width, 100%) * ${x})) translateY(calc(var(--motion-height, 100%) * ${y})) scale(1) translateX(calc(var(--motion-width, 100%) * ${-x})) translateY(calc(var(--motion-height, 100%) * ${-y})) rotate(var(--comp-rotate-z, 0deg))`
    }]
  }];

  /*
   *
   * @keyframes <name>-scale {
   *   from {
   *    transform: `translateX(calc(var(--motion-width, 100%) * <x>)) translateY(calc(var(--motion-height, 100%) * <y>)) scale(<scale>) translateX(calc(var(--motion-width, 100%) * <-x>)) translateY(calc(var(--motion-height, 100%) * <-y>))  rotate(var(--comp-rotate-z, 0deg))`;
   *   }
   *   to {
   *    transform: `translateX(calc(var(--motion-width, 100%) * <x>)) translateY(calc(var(--motion-height, 100%) * <y>)) scale(1) translateX(calc(var(--motion-width, 100%) * <-x>)) translateY(calc(var(--motion-height, 100%) * <-y>)) rotate(var(--comp-rotate-z, 0deg))`;
   *   }
   * }
   *
   * @keyframes <name>-fade {
   *   from {
   *     opacity: 0;
   *   }
   *   to {
   *     opacity: 1;
   *   }
   * }
   *
   * #target {
   *   animation: <name>-scale <duration> <delay> <easing> backwards,
   *              <name>-fade <duration> <delay> linear backwards;
   * }
   */
}
