// @wix/ambassador-pro-gallery-v2-gallery/build/es/http.impl.js
import { toURLSearchParams } from '@wix/metro-runtime';
import { serializer } from '@wix/metro-runtime/ambassador';
import { resolveUrl } from '@wix/metro-runtime';
var _bulkDeleteGalleryItemsRequest = {};
var _bulkDeleteGalleryItemsResponse = {};
var _commonImage = { urlExpirationDate: 'google.protobuf.Timestamp' };
var _createGalleryItemRequest = { item: '_item' };
var _createGalleryItemResponse = { item: '_item' };
var _createGalleryRequest = { gallery: '_gallery' };
var _createGalleryResponse = { gallery: '_gallery' };
var _deleteGalleryItemRequest = {};
var _deleteGalleryItemResponse = {};
var _deleteGalleryItemsRequest = {};
var _deleteGalleryItemsResponse = { gallery: '_gallery' };
var _deleteGalleryRequest = {};
var _deleteGalleryResponse = {};
var _gallery = {
    sortOrder: 'google.protobuf.DoubleValue',
    createdDate: 'google.protobuf.Timestamp',
    items: '_item',
};
var _getGalleryItemRequest = {};
var _getGalleryItemResponse = { item: '_item' };
var _getGalleryRequest = {};
var _getGalleryResponse = { gallery: '_gallery' };
var _image = {
    imageInfo: '_commonImage',
    focalPoint: '_point',
    unsharpMasking: '_unsharpMasking',
};
var _item = {
    sortOrder: 'google.protobuf.DoubleValue',
    createdDate: 'google.protobuf.Timestamp',
    updatedDate: 'google.protobuf.Timestamp',
    image: '_image',
    video: '_video',
    secondaryMedia: '_secondaryMedia',
};
var _listGalleriesRequest = { fields: 'google.protobuf.FieldMask' };
var _listGalleriesResponse = { galleries: '_gallery' };
var _listGalleryItemsRequest = {};
var _listGalleryItemsResponse = { items: '_item' };
var _point = { x: 'DOUBLE', y: 'DOUBLE' };
var _publishGalleryRequest = {};
var _publishGalleryResponse = { gallery: '_gallery' };
var _secondaryMedia = { image: '_image' };
var _unsharpMasking = {
    amount: 'google.protobuf.FloatValue',
    radius: 'google.protobuf.FloatValue',
    threshold: 'google.protobuf.FloatValue',
};
var _updateGalleryItemRequest = { item: '_item' };
var _updateGalleryItemResponse = { item: '_item' };
var _updateGalleryRequest = { gallery: '_gallery' };
var _updateGalleryResponse = { gallery: '_gallery' };
var _video = { videoInfo: '_videoV2' };
var _videoResolution = { poster: '_commonImage' };
var _videoV2 = {
    urlExpirationDate: 'google.protobuf.Timestamp',
    resolutions: '_videoResolution',
    posters: '_commonImage',
};
function resolveComWixpressExposureProgalleryProGalleryUrl(opts) {
    var domainToMappings = {
        _: [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'manage._base_domain_': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'www._base_domain_': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'editor._base_domain_': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'blocks._base_domain_': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'create.editorx': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'progallery.wixapps.net': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
        'www.wixapis.com': [
            {
                srcPath: '/progallery',
                destPath: '',
            },
        ],
        'api._api_base_domain_': [
            {
                srcPath: '/pro-gallery-server',
                destPath: '',
            },
        ],
        '*.dev.wix-code.com': [
            {
                srcPath: '/_api/pro-gallery-server',
                destPath: '',
            },
        ],
    };
    return resolveUrl(Object.assign(opts, { domainToMappings: domainToMappings }));
}
/** Retrieves a list of galleries. */
export function listGalleries(payload) {
    var _a = serializer(_listGalleriesRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_listGalleriesResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __listGalleries(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'GET',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.ListGalleries',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __listGalleries.fromReq = fromReq;
    __listGalleries.__isAmbassador = true;
    return __listGalleries;
}
/** Gets a gallery by ID. */
export function getGallery(payload) {
    var _a = serializer(_getGalleryRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_getGalleryResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __getGallery(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'GET',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.GetGallery',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __getGallery.fromReq = fromReq;
    __getGallery.__isAmbassador = true;
    return __getGallery;
}
/** Retrieves a list of media items in a specified gallery. */
export function listGalleryItems(payload) {
    var _a = serializer(_listGalleryItemsRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_listGalleryItemsResponse, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __listGalleryItems(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'GET',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.ListGalleryItems',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __listGalleryItems.fromReq = fromReq;
    __listGalleryItems.__isAmbassador = true;
    return __listGalleryItems;
}
/** Gets a gallery item by ID. */
export function getGalleryItem(payload) {
    var _a = serializer(_getGalleryItemRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_getGalleryItemResponse, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __getGalleryItem(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'GET',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.GetGalleryItem',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items/{itemId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __getGalleryItem.fromReq = fromReq;
    __getGalleryItem.__isAmbassador = true;
    return __getGalleryItem;
}
/**
 * Creates a new gallery.
 *
 * > **Note:** When you create a new gallery with Create Gallery, the newly-created gallery is available only on a site's backend and doesn't appear on the live site.
 * > To display a backend gallery on a live site, the site owner needs to connect it manually to a gallery component.
 * > To learn more about how a site owner can do this using Velo, see [Displaying a Pro Gallery on Your Site Using the Pro Gallery Backend API](https://support.wix.com/en/article/velo-tutorial-displaying-a-pro-gallery-on-your-site-using-the-pro-gallery-backend-api).
 *
 * You can create an entirely new gallery, or clone an existing gallery using the `cloneFromGalleryId` parameter.
 * When a gallery is cloned, the newly-created gallery includes the same properties as the existing gallery except for the gallery IDs, item IDs, sort order, and created and updated dates.
 */
export function createGallery(payload) {
    var _a = serializer(_createGalleryRequest, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_createGalleryResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __createGallery(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'POST',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.CreateGallery',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __createGallery.fromReq = fromReq;
    __createGallery.__isAmbassador = true;
    return __createGallery;
}
/** Updates a gallery. */
export function updateGallery(payload) {
    var _a = serializer(_updateGalleryRequest, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_updateGalleryResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __updateGallery(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'PATCH',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.UpdateGallery',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{gallery.id}',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __updateGallery.fromReq = fromReq;
    __updateGallery.__isAmbassador = true;
    return __updateGallery;
}
/**
 * Deletes a gallery.
 *
 * >**Notes:**
 * > * The deleted gallery appears as an empty widget on the live site.
 * > * The deleted gallery is no longer retrieved when calling the [List Galleries](https://dev.wix.com/api/rest/site-content/pro-gallery/list-galleries) endpoint.
 * >
 */
export function deleteGallery(payload) {
    var _a = serializer(_deleteGalleryRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_deleteGalleryResponse, {}).fromJSON;
    function __deleteGallery(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'DELETE',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.DeleteGallery',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __deleteGallery.fromReq = fromReq;
    __deleteGallery.__isAmbassador = true;
    return __deleteGallery;
}
/**
 * Deletes multiple media items from a gallery.
 * @deprecated It has been replaced with com.wixpress.exposure.progallery.BulkDeleteGalleryItems(), and will be removed on 2024-08-15.
 */
export function deleteGalleryItems(payload) {
    var _a = serializer(_deleteGalleryItemsRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_deleteGalleryItemsResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __deleteGalleryItems(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'DELETE',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.DeleteGalleryItems',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __deleteGalleryItems.fromReq = fromReq;
    __deleteGalleryItems.__isAmbassador = true;
    return __deleteGalleryItems;
}
/** Deletes multiple media items from a gallery. */
export function bulkDeleteGalleryItems(payload) {
    var _a = serializer(_bulkDeleteGalleryItemsRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_bulkDeleteGalleryItemsResponse, {}).fromJSON;
    function __bulkDeleteGalleryItems(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'POST',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.BulkDeleteGalleryItems',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items/delete',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __bulkDeleteGalleryItems.fromReq = fromReq;
    __bulkDeleteGalleryItems.__isAmbassador = true;
    return __bulkDeleteGalleryItems;
}
/**
 * Creates a media item in a specified gallery.
 *
 * > **Note:** When creating or updating image items in a pro gallery, the images must be uploaded or imported to the [Wix Media Manager](https://dev.wix.com/api/rest/media/media-manager) first, as the `imageInfo` parameter currently only supports the Wix Media URL.
 */
export function createGalleryItem(payload) {
    var _a = serializer(_createGalleryItemRequest, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_createGalleryItemResponse, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __createGalleryItem(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'POST',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.CreateGalleryItem',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __createGalleryItem.fromReq = fromReq;
    __createGalleryItem.__isAmbassador = true;
    return __createGalleryItem;
}
/**
 * Updates a media item in a specified gallery.
 *
 * > **Note:** When creating or updating image items in a pro gallery, the images must be uploaded or imported to the [Wix Media Manager](https://dev.wix.com/api/rest/media/media-manager) first, as the `imageInfo` parameter currently only supports the Wix Media URL.
 */
export function updateGalleryItem(payload) {
    var _a = serializer(_updateGalleryItemRequest, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_updateGalleryItemResponse, {
        _commonImage: _commonImage,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __updateGalleryItem(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'PATCH',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.UpdateGalleryItem',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items/{item.id}',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __updateGalleryItem.fromReq = fromReq;
    __updateGalleryItem.__isAmbassador = true;
    return __updateGalleryItem;
}
/** Deletes a media item from a gallery. */
export function deleteGalleryItem(payload) {
    var _a = serializer(_deleteGalleryItemRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_deleteGalleryItemResponse, {}).fromJSON;
    function __deleteGalleryItem(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'DELETE',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.DeleteGalleryItem',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/items/{itemId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __deleteGalleryItem.fromReq = fromReq;
    __deleteGalleryItem.__isAmbassador = true;
    return __deleteGalleryItem;
}
/**
 * Publish gallery
 * The changes that been done in SAVED state will be applied in the PUBLISHED gallery
 */
export function publishGallery(payload) {
    var _a = serializer(_publishGalleryRequest, {}), toReq = _a.toJSON, fromReq = _a.fromJSON;
    var fromRes = serializer(_publishGalleryResponse, {
        _commonImage: _commonImage,
        _gallery: _gallery,
        _image: _image,
        _item: _item,
        _point: _point,
        _secondaryMedia: _secondaryMedia,
        _unsharpMasking: _unsharpMasking,
        _video: _video,
        _videoResolution: _videoResolution,
        _videoV2: _videoV2,
    }).fromJSON;
    function __publishGallery(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.pro_gallery.v2.gallery',
            method: 'POST',
            methodFqn: 'com.wixpress.exposure.progallery.ProGallery.PublishGallery',
            url: resolveComWixpressExposureProgalleryProGalleryUrl({
                protoPath: '/v2/galleries/{galleryId}/publish',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __publishGallery.fromReq = fromReq;
    __publishGallery.__isAmbassador = true;
    return __publishGallery;
}
