// @wix/ambassador-pro-gallery-v2-gallery/build/es/types.impl.js
export var Type;
(function (Type) {
    Type["UNDEFINED"] = "UNDEFINED";
    /** external link */
    Type["EXTERNAL"] = "EXTERNAL";
    /** for internal usage using wixLinkData */
    Type["INTERNAL"] = "INTERNAL";
})(Type || (Type = {}));
/**
 * The 'rel' attribute of the link. The rel attribute defines the relationship between a linked resource and the current document.
 * Further reading (also about different possible rel types): https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel
 * Following are the accepted 'rel' types by Wix applications.
 */
export var LinkRel;
(function (LinkRel) {
    /** default (not implemented) */
    LinkRel["unknown_link_rel"] = "unknown_link_rel";
    /** Indicates that the current document's original author or publisher does not endorse the referenced document. */
    LinkRel["nofollow"] = "nofollow";
    /** Instructs the browser to navigate to the target resource without granting the new browsing context access to the document that opened it. */
    LinkRel["noopener"] = "noopener";
    /** No Referer header will be included. Additionally, has the same effect as noopener. */
    LinkRel["noreferrer"] = "noreferrer";
    /** Indicates a link that resulted from advertisements or paid placements. */
    LinkRel["sponsored"] = "sponsored";
})(LinkRel || (LinkRel = {}));
export var ItemType;
(function (ItemType) {
    ItemType["UNDEFINED"] = "UNDEFINED";
    ItemType["IMAGE"] = "IMAGE";
    ItemType["VIDEO"] = "VIDEO";
    ItemType["TEXT"] = "TEXT";
})(ItemType || (ItemType = {}));
export var ImageType;
(function (ImageType) {
    ImageType["UNDEFINED"] = "UNDEFINED";
    ImageType["WIX_MEDIA"] = "WIX_MEDIA";
    ImageType["EXTERNAL"] = "EXTERNAL";
})(ImageType || (ImageType = {}));
export var VideoType;
(function (VideoType) {
    VideoType["UNDEFINED"] = "UNDEFINED";
    VideoType["WIX_MEDIA"] = "WIX_MEDIA";
    VideoType["YOUTUBE"] = "YOUTUBE";
    VideoType["VIMEO"] = "VIMEO";
})(VideoType || (VideoType = {}));
export var WebhookIdentityType;
(function (WebhookIdentityType) {
    WebhookIdentityType["UNKNOWN"] = "UNKNOWN";
    WebhookIdentityType["ANONYMOUS_VISITOR"] = "ANONYMOUS_VISITOR";
    WebhookIdentityType["MEMBER"] = "MEMBER";
    WebhookIdentityType["WIX_USER"] = "WIX_USER";
    WebhookIdentityType["APP"] = "APP";
})(WebhookIdentityType || (WebhookIdentityType = {}));
export var Enum;
(function (Enum) {
    /** Default value. Means that permission not set */
    Enum["UNKNOWN"] = "UNKNOWN";
    /** Protected exposure. Exposed to members and owners */
    Enum["PROTECTED"] = "PROTECTED";
    /** Private exposure. Exposed to owners */
    Enum["PRIVATE"] = "PRIVATE";
    /** Public exposure. Visible to everyone */
    Enum["PUBLIC"] = "PUBLIC";
    /** Used for partial updates, to state that exposure is not changing */
    Enum["UNCHANGED"] = "UNCHANGED";
    /** Protected to members of permitted groups and owners */
    Enum["GROUP_PROTECTED"] = "GROUP_PROTECTED";
})(Enum || (Enum = {}));
export var SearchIndexingNotificationState;
(function (SearchIndexingNotificationState) {
    /** default state */
    SearchIndexingNotificationState["Unknown"] = "Unknown";
    /** metasite does not require site search indexing */
    SearchIndexingNotificationState["Off"] = "Off";
    /** metasite requires site search indexing */
    SearchIndexingNotificationState["On"] = "On";
})(SearchIndexingNotificationState || (SearchIndexingNotificationState = {}));
export var State;
(function (State) {
    State["UNDEFINED"] = "UNDEFINED";
    /** The gallery in the Editor segment */
    State["SAVED"] = "SAVED";
    /** The gallery in the LiveSite segment */
    State["PUBLISHED"] = "PUBLISHED";
})(State || (State = {}));
