// @wix/error-handler/dist/esm/error-handler/reportErrorResolutionBi.js
import { httpClientErrorMappingSrc176Evid100 } from '@wix/bi-logger-os-data/v2';
import { withoutDefaults } from '../util/withoutDefaults';
export function reportErrorResolutionBi(errorHandler, params) {
  var _errorCodesMap$applic, _errorToShow$action;
  const {
    errorAccessor,
    errorResolution,
    errorCodesMap
  } = params;
  const {
    biLogger
  } = errorHandler;
  const {
    error,
    applicationError
  } = errorAccessor;
  const {
    errorToShow,
    consumerResolvedError
  } = errorResolution ?? {};
  const code = applicationError == null ? void 0 : applicationError.code;
  const applicationErrorMapping = errorCodesMap == null || (_errorCodesMap$applic = errorCodesMap.applicationError) == null ? void 0 : _errorCodesMap$applic[code];
  biLogger.report(withoutDefaults(httpClientErrorMappingSrc176Evid100)({
    error_opt_out: applicationErrorMapping === false,
    translated_message: errorToShow == null ? void 0 : errorToShow.message,
    action_text: errorToShow == null || (_errorToShow$action = errorToShow.action) == null ? void 0 : _errorToShow$action.text,
    is_override_error: String(errorCodesMap != null),
    is_consumer_resolved_error: consumerResolvedError != null,
    session_id: error == null ? void 0 : error._errorHandlerSessionId
  }));
}
