// @wix/error-handler/dist/esm/state/commonErrorsMapping.js
export async function commonErrorsMapping(params) {
  const {
    getTranslateFn
  } = params;
  const t = await getTranslateFn();
  return {
    applicationError: {
      CAPTCHA_REQUIRED: () => ({
        message: t('errorHandler.common-error.captchaRequired'),
        action: {
          text: t('errorHandler.common-error.captchaRequired.verify.cta')
        }
      })
    },
    statusCodeError: {
      UNAUTHENTICATED: () => ({
        message: t('errorHandler.common-error.unauthenticated'),
        action: {
          text: t('errorHandler.common-error.unauthenticated.cta')
        }
      }),
      INVALID_ARGUMENT: () => ({
        message: t('errorHandler.common-error.validationBasic')
      }),
      PERMISSION_DENIED: () => ({
        message: t('errorHandler.common-error.permission-denied.site')
      }),
      NOT_FOUND: () => ({
        message: t('errorHandler.common-error.not-found'),
        action: {
          text: t('errorHandler.common-error.tryAgain.cta')
        }
      }),
      ALREADY_EXISTS: () => ({
        message: t('errorHandler.common-error.conflict'),
        action: {
          text: t('errorHandler.common-error.refresh.cta')
        }
      }),
      UNAVAILABLE: () => ({
        message: t('errorHandler.common-error.technicalIssue')
      }),
      RESOURCE_EXHAUSTED: () => ({
        message: t('errorHandler.common-error.resource-exhausted')
      }),
      BAD_GATEWAY: () => ({
        message: t('errorHandler.common-error.technicalIssue'),
        action: {
          text: t('errorHandler.common-error.tryAgain.cta')
        }
      }),
      GATEWAY_TIMEOUT: () => ({
        message: t('errorHandler.common-error.technicalIssue')
      }),
      INTERNAL: () => ({
        message: t('errorHandler.common-error.technicalIssue')
      })
    }
  };
}
export async function connectionError(params) {
  const {
    getTranslateFn
  } = params;
  const t = await getTranslateFn();
  return {
    message: t('errorHandler.common-error.noConnection'),
    action: {
      text: t('errorHandler.common-error.tryAgain.cta')
    }
  };
}
export async function defaultError(params) {
  const {
    getTranslateFn
  } = params;
  const t = await getTranslateFn();
  return {
    message: t('errorHandler.common-error.technicalIssue')
  };
}
