// @wix/error-handler/dist/esm/state/ErrorResolution.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { resolveFromMapping } from '../util/resolveFromMapping';
import { defaultError } from './commonErrorsMapping';
const errorCodesMapPromise = import( /* webpackChunkName: "error-handler-commonErrorsMapping" */'./commonErrorsMapping');
export class ErrorResolution {
  constructor(params) {
    _defineProperty(this, "errorAccessor", void 0);
    _defineProperty(this, "errorCodesMap", void 0);
    _defineProperty(this, "getTranslateFn", void 0);
    _defineProperty(this, "onlineManager", void 0);
    _defineProperty(this, "consumerResolvedError", void 0);
    _defineProperty(this, "commonResolvedError", void 0);
    this.getTranslateFn = params.getTranslateFn;
    this.errorAccessor = params.errorAccessor;
    this.errorCodesMap = params.errorCodesMap;
    this.onlineManager = params.onlineManager;
  }
  async _resolveNetworkError() {
    const {
      onlineManager
    } = this;
    const {
      connectionError
    } = await errorCodesMapPromise;
    if (onlineManager.onLine) {
      return defaultError({
        getTranslateFn: this.getTranslateFn
      });
    }
    return connectionError(this);
  }
  async _resolveCommonError() {
    const {
      errorAccessor
    } = this;
    const {
      httpError
    } = errorAccessor;
    const {
      commonErrorsMapping
    } = await errorCodesMapPromise;
    if (httpError == null) {
      return this._resolveNetworkError();
    }
    const resolvedError = resolveFromMapping({
      errorAccessor,
      httpError,
      errorCodesMap: await commonErrorsMapping(this)
    });
    if (!resolvedError) {
      return defaultError({
        getTranslateFn: this.getTranslateFn
      });
    }
    return resolvedError;
  }
  _resolveConsumerError() {
    const {
      errorCodesMap,
      errorAccessor
    } = this;
    const {
      httpError
    } = errorAccessor;
    if (httpError == null || errorCodesMap == null) {
      return null;
    }
    return resolveFromMapping({
      errorAccessor,
      errorCodesMap,
      httpError
    });
  }
  async resolve() {
    this.consumerResolvedError = this._resolveConsumerError();

    // If consumer resolved the error, we don't need to resolve the common error
    if (this.consumerResolvedError != null) {
      return;
    }
    this.commonResolvedError = await this._resolveCommonError();
  }
  get errorToShow() {
    return this.consumerResolvedError ?? this.commonResolvedError;
  }
}
